sqlmesh/core/macros.py

```diff
@@ -349,6 +354,65 @@ def engine_adapter(self) -> EngineAdapter:
             )
         return self.locals["engine_adapter"]
 
+    def _coerce(self, expr: exp.Expression, typ: t.Type) -> t.Any:
+        """Coerces the given expression to the specified type on a best-effort basis."""
+        try:
+            if typ is None:
+                return expr
+            base = t.get_origin(typ) or typ
+            if isinstance(expr, base):
+                return expr
+            if t.get_origin(typ) is t.Union:
+                for t_ in t.get_args(typ):
+                    try:
+                        return self._coerce(expr, t_)
+                    except Exception:
+                        pass
+                raise SQLMeshError(f"Failed to coerce expression '{expr}' to type '{typ}'.")
```

Comment:
this raise is repeated several times, could we make this more dry?

Reply:
Sure